### ISSUE TITLE
Possibility to overnight writer methods

### ISSUE THREAD

**Issue by lafriks** [niklasfasching/go-org#14]
Currently it is impossible to override HTML writer methods to fine-tune them. For Gitea project we need custom link/image URLs but only way to implement this is to copy all HTML writer and change single method. Mostly it is because of nodeAsString method that clones writer and also WriteNode usage that passes HTML writer all over the place

**Comment by lafriks** [comment]
For good example you can check blackfriday v2 renderer interface/HTML writer implementation that is easy to customize for specific needs

**Comment by niklasfasching** [comment]
I'd love to do that. I actually tried to do it with the current interface because i was mistaken about how embedding works and thought it to be like inheritance, see [here](https://github.com/niklasfasching/go-org/commit/faea88d48e957ad97a1ddaf781ae86c6f251931b) (note the comment below the diff).

At a glance I don't see how blackfriday v2 makes this possible, could you elaborate? In any case I'll look into this next week and hope blackfriday presents an elegant solution that we can imitate.

**Comment by lafriks** [comment]
Check this PR in Gitea how we are overriding it for blackfriday: 
https://github.com/go-gitea/gitea/pull/8560

**Comment by niklasfasching** [comment]
Thx! I missed [WithRenderer](https://godoc.org/gopkg.in/russross/blackfriday.v2#WithRenderer) while skimming the code yesterday. Looks simple and clean enough. Will add it end of next week :)

**Comment by niklasfasching** [comment]
Implemented in [14900e9](https://github.com/niklasfasching/go-org/commit/14900e97e20d41070e4d9a2918857e9b4e1c4032). 
Usage: 
```
type ExtendedHTMLWriter struct { *HTMLWriter }
func (w *ExtendedHTMLWriter) WriteText(t Text) {}

func NexExtendedHTMLWriter() *ExtendedHTMLWriter {
	htmlWriter := NewHTMLWriter()
	extendedWriter := &ExtendedHTMLWriter{htmlWriter, 0}
	htmlWriter.ExtendingWriter = extendedWriter
	return extendedWriter
}
```

**Comment by niklasfasching** [comment]
feel free to re-open this if you have anything - I just like to keep my inbox empty so I'm closing this for now.

**Comment by lafriks** [comment]
Yes, it does work, thanks. It is still tricky with needing own `nodesAsString` and `emptyClone` functions but otherwise it does work great. Thanks again!

**Comment by niklasfasching** [comment]
Could you point me to the part where you run into this problem? I thought about getting rid of cloning but in the end thought it wasn't needed. In any case the code is still there and maybe this solves your problems? https://github.com/niklasfasching/go-org/commit/f0ebfce576fee6827dc1e5aef27bed0953159388

**Comment by lafriks** [comment]
I need them because I'm overriding WriteRegularLink function, see:

https://github.com/go-gitea/gitea/pull/8560/files#diff-ec1158a92d009b3ac2d4241bac8c8e63R92

**Comment by niklasfasching** [comment]
Don't have time to look into this tonight but sounds like this would be solved by exporting `nodesAsString` (which already forwards to the `extendingWriter`). How about https://github.com/niklasfasching/go-org/commit/3c380c814e46c76c42ffc03a5f1f36505680a332 ?

**Comment by lafriks** [comment]
It does not work correct with that commit, it renders text that should be between `<a>`tags twice:

        	            	expected: "<p>\n<a href=\"http://localhost:3000/gogits/gogs/WikiPage\" title=\"WikiPage\">WikiPage</a>\n</p>"
        	            	actual  : "<p>\nWikiPage<a href=\"http://localhost:3000/gogits/gogs/WikiPage\" title=\"\"></a>\n</p>"

**Comment by lafriks** [comment]
I had to make clone function this way to get it working for me:
```go
func (r *Renderer) emptyClone() *Renderer {
	wcopy := *(r.HTMLWriter)
	wcopy.Builder = strings.Builder{}

	rcopy := *r
	rcopy.HTMLWriter = &wcopy

	wcopy.ExtendingWriter = &rcopy

	return &rcopy
}

func (r *Renderer) nodesAsString(nodes ...org.Node) string {
	tmp := r.emptyClone()
	org.WriteNodes(tmp, nodes...)
	return tmp.String()
}
```

**Comment by niklasfasching** [comment]
Can you try again? https://github.com/niklasfasching/go-org/pull/15
I remembered why I removed cloning :sweat_smile: - without it it works afaict.

**Comment by niklasfasching** [comment]
Thx for testing, merged in #15 (v0.1.8)

**Comment by lafriks** [comment]
Thanks
